contracts/base-chain/bridge-gateways/matic/AMPLMaticTransferGateway.sol

```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.7.3;
+
+import {FxBaseRootTunnel} from "fx-portal/contracts/tunnel/FxBaseRootTunnel.sol";
+import {SafeMath} from "@openzeppelin/contracts/math/SafeMath.sol";
+
+import {ILayer2BridgeGateway} from "../../../_interfaces/ILayer2BridgeGateway.sol";
+import {ITokenVault} from "../../../_interfaces/ITokenVault.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+/**
+ * @title AMPLMaticTransferGateway: AMPL-Matic Transfer Gateway Contract
+ * @dev This contract is deployed on the base chain (Ethereum).
+ *
+ *      It's a pass-through contract between the Matic's bridge contracts and
+ *      the and the Token vault.
+ *
+ */
+contract AMPLMaticTransferGateway is ILayer2BridgeGateway, FxBaseRootTunnel {
+    using SafeMath for uint256;
+
+    address public immutable ampl;
+    address public immutable vault;
+
+    /**
+     * @dev Calculates the amount of amples to be unlocked based on the share of total supply and
+     *      transfers it to the recipient.
+     *      "senderAddressInSourceChain": Address of the sender wallet in the matic chain.
+     *      "recipient": Address of the recipient wallet in ethereum.
+     *      "amount": Amount of tokens that were {burnt} on the matic.
+     *      "globalAMPLSupply": AMPL ERC-20 total supply at the time of transfer.
+     */
+    function _processMessageFromChild(bytes memory data) internal override {
+        address senderInSourceChain;
+        address recipient;
+        uint256 amount;
+        uint256 globalAMPLSupply;
+        (senderInSourceChain, recipient, amount, globalAMPLSupply) = abi.decode(
```

Comment:
Do we need to do any checks related to source chain?

For example, make sure it's impossible to send from a testnet to a mainnet?

Reply:
Once these contracts are deployed, they need to be configured to point to the correct partner contract on the other chain.

eg) `AMPLMaticRebaseGateway` on ethereum stores the reference to the `MaticXCAmpleRebaseGateway` on matic through the `setFxChildTunnel` which is setup after deployment (which can only be set once) and similarly `MaticXCAmpleRebaseGateway` on matic stores the reference to `AMPLMaticRebaseGateway` on ethereum.

https://docs.matic.network/docs/develop/l1-l2-communication/state-transfer